pkg/kvstore/store/syncstore.go

```diff
@@ -0,0 +1,190 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package store
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"path"
+	"sync"
+
+	"github.com/sirupsen/logrus"
+	"k8s.io/client-go/util/workqueue"
+
+	"github.com/cilium/cilium/pkg/logging/logfields"
+)
+
+// SyncStore abstracts the operations allowing to synchronize key/value pairs
+// into a kvstore.
+type SyncStore interface {
+	// UpsertKey upserts a key/value pair into the kvstore.
+	UpsertKey(ctx context.Context, key Key) error
+
+	// DeleteLocalKey removes a key from the kvstore.
+	DeleteKey(ctx context.Context, key NamedKey) error
+
+	// Release frees all resources owned by the store, leaving all keys in the
+	// kvstore intact.
+	Release()
+}
+
+// SyncStoreBackend represents the subset kvstore.BackendOperations leveraged
+// by SyncStore implementations.
+type SyncStoreBackend interface {
+	// Update creates or updates a key.
+	Update(ctx context.Context, key string, value []byte, lease bool) error
+	// Delete deletes a key.
+	Delete(ctx context.Context, key string) error
+}
+
+// wqSyncStore implements the SyncStore interface leveraging a workqueue to
+// coalescence update/delete requests and handle retries in case of errors.
+type wqSyncStore struct {
+	backend   SyncStoreBackend
+	prefix    string
+	withLease bool
+
+	ctx    context.Context
+	cancel context.CancelFunc
+
+	workqueue workqueue.RateLimitingInterface
+	state     sync.Map /* map[string][]byte --- map[NamedKey.GetKeyName()]Key.Marhal() */
+
+	log *logrus.Entry
+	wg  sync.WaitGroup
+}
+
+// NewWorkqueueSyncStore returns a SyncStore instance which leverages a workqueue
+// to coalescence update/delete requests and handle retries in case of errors.
+func NewWorkqueueSyncStore(backend SyncStoreBackend, prefix string) SyncStore {
+	w := &wqSyncStore{
+		backend:   backend,
+		prefix:    prefix,
+		withLease: true,
+
+		workqueue: workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+
+		log: log.WithField("prefix", prefix),
+	}
+
+	w.ctx, w.cancel = context.WithCancel(context.Background())
+
+	w.wg.Add(1)
+	go w.syncLoop()
```

Comment:
I would like to propose a slightly cleaner interface. 
First, we can remove ctx from `wqSyncStore` struct.
We could also add method, something like:
```
func (w *wqSyncStore) Run(ctx context.Context, workers int) {
	defer w.workqueue.ShutDown()

	log.Infof("Starting wgSyncStore")
	defer log.Infof("Shutting down wgSyncStore")

	for i := 0; i < workers; i++ {
		go w.syncLoop(ctx)
	}

	<-ctx.Done()
}
```

Reply:
Yep, makes definitely sense to me. I've just slightly changed it to remove the worker parameter (configured at creation time) to generalize the interface, and added a waitgroup to wait for all goroutines to terminate before returning.